src/v/coproc/wasm_event_listener.cc

```diff
@@ -0,0 +1,216 @@
+                "Moving coprocessor {} to inactive dir",
+                active_path);
+              return ss::rename_file(active_path.string(), mvto.string());
+          } else {
+              if (exists) {
+                  vlog(
+                    coproclog.warn,
+                    "Attempt to deploy coprocessor with script {} but one "
+                    "already exists: {}",
+                    name,
+                    active_path);
+                  return ss::now();
+              }
+              vlog(
+                coproclog.info,
+                "Writing coprocessor {} to submit dir {}",
+                name,
+                _submit_dir);
+              std::filesystem::path submit_src(_submit_dir / name.c_str());
+              return write_fully(submit_src, std::move(source_code));
+          }
+      });
+}
+
+wasm_event_listener::wasm_event_listener(std::filesystem::path wasm_root)
+  : _client(
+    std::vector<unresolved_address>{unresolved_address("127.0.0.1", 9092)})
+  , _wasm_root(std::move(wasm_root))
+  , _active_dir(_wasm_root / "active")
+  , _submit_dir(_wasm_root / "submit")
+  , _inactive_dir(_wasm_root / "inactive") {}
+
+void wasm_event_listener::start() {
+    if (_gate.is_closed() || _gate.get_count() != 0) {
+        throw std::logic_error(
+          "Attempted to start() the wasm_event_notifier run "
+          "loop, after it has already been started or closed");
+    }
+    (void)ss::with_gate(_gate, [this] {
+        return storage::directories::initialize(_active_dir.string())
+          .then([this] {
+              return storage::directories::initialize(_inactive_dir.string());
+          })
+          .then([this] {
+              return storage::directories::initialize(_submit_dir.string());
+          })
+          .then([this] { return _client.connect(); })
+          .then([this] {
+              return ss::do_until(
+                [this] { return _abort_source.abort_requested(); },
+                [this] { return do_start(); });
+          });
+    });
+}
+
+ss::future<> wasm_event_listener::do_start() {
+    /// This method performs the main polling behavior. Within a repeat loop it
+    /// will poll from data until it cannot poll anymore. Normally we would be
+    /// concerned about keeping all of this data in memory, however the topic is
+    /// compacted, we don't expect the size of unique records to be very big.
+    return ss::do_with(
+      model::record_batch_reader::data_t(),
+      [this](model::record_batch_reader::data_t& events) {
+          /// The stop condition is met when its been detected that the stored
+          /// offset has not moved.
+          return ss::repeat([this, &events] {
+                     return poll_topic(events).then([this, o = _offset] {
+                         return o == _offset ? ss::stop_iteration::yes
+                                             : ss::stop_iteration::no;
+                     });
```

Comment:
Maybe a named temporary prior to the capture would make this a little less subtle.
```cpp
                     auto initial_offset = _offset;
	             return poll_topic(events).then([this, initial_offset] {
                         return initial_offset == _offset ? ss::stop_iteration::yes
                                                          : ss::stop_iteration::no;
                     });
```